### ISSUE TITLE
Nearby: invalidate cache after uploading from Nearby

### ISSUE THREAD

**Issue by nicolas-raoul** [commons-app/apps-android-commons#5808]
The pin does not turn from red to green after upload.

Steps to reproduce:
1. Open Nearby
2. Select a red pin
3. Upload a picture for it
4. Close and reopen the app, open Nearby again
5. Problem: The pin is still red. It looks like I need to upload again.

Fix: The app should remove that pin's cache item after the upload.

If easier, removing all pins from the cache is OK.

**Comment by ChristoJobyAntony** [comment]
Could I have a go at this ?

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul, Once again, thank you for the assignment. I have been digging around a bit to understand how the places are being cached. My observations were the following:
1. The classes `PlacesRepository`,  `PlacesLocalDataSource` & `PlaceDao` implement local caching of the places fetched from Wiki. Currently, these classes only implement an asynchronous and synchronous insertion logic. 
2. The `PlaceRepository` is being used in `NearbyParentFragment` it has been injected with the instance. 

My proposal on solving the current solution:
1.  Add a delete method to the repository to delete a given place.
2.  Add the  `PlacesRepository` to the `UploadActivity` and inject it.
3.  On upload complete, use the `PlacesRepository` to delete the place.

I would love your feedback on this pathway of implementation.

**Comment by nicolas-raoul** [comment]
Sounds good! :-)

By the way https://github.com/commons-app/apps-android-commons/issues/5806 could use a similar approach.

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul  I have been working on this fix in the Beta version of the application. I wonder if an image depiction in the Beta server is enough to mark a pin green. Because, despite multiple uploads, I am not able to make a pin turn green.

**Comment by nicolas-raoul** [comment]
Some features do not work on the beta server indeed. In such cases, you can use the production server with care, for instance using sandbox element https://www.wikidata.org/wiki/Q4115189 , not too many times.

Do not hesitate to modify the sandbox element's geographical location to somewhere convenient for you, and remove its P18 image.

Please only upload encyclopedia-worthy images to the production server, I can send you good images to upload if you want _(in that case please send me an email to my address which can be found at https://nicolasraoul.github.io)_.

Thanks! :-)

**Comment by ChristoJobyAntony** [comment]
Thank you for the hint.  I have modified the location of the sandbox item to a more accessible location. 
Would it be fine to upload this picture of ducklings ?
![WhatsApp Image 2024-10-16 at 23 40 03_de45d045](https://github.com/user-attachments/assets/d81f17b1-683c-41d6-bb57-989079fae689)

**Comment by nicolas-raoul** [comment]
Yes such pictures are fine if you include geolocation. Clear pictures of buildings and food are also great, especially if you include the name of the building or dish. Thanks!

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul A concern with this approach. I am worried that if I clear a cache immediately on image submit (before the upload is complete), the Nearby query to fetch items may still find the depiction of the image. Do you think I should clear the cache once the upload completes ?

**Comment by nicolas-raoul** [comment]
Yes, clear cache only once the P18 association has been made (I think there is a toast saying "image added to Wikidata" or something).

**Comment by ChristoJobyAntony** [comment]
@nicolas-raoul, thank you for the guidance here. I dived a bit more onto the codebase to find the chain of calls. That handled the association. I was able to track down the request to edit `WikiItems` as the following trail of method calls:
1. `UploadWorker.makeWikiEdit`  verifies if a contribution has a valid location and can be claimed to be depicting a `WikiItem`
2. If the `Contribution` is deemed to be valid, then a claim is created through the `wikidataEditServices.createClaim`.
3. On a successful claim, a new `revisionId` is returned, and the `wikidataEditService.handleImageClaimResult` is dispatched in a separate thread with the Main Context. 

From this observation, I initially attempted to inject my code to delete cached `WikiItem` in the `handleImageClaimResult` handler, since it seemed to be most appropriate. However, after a lot of debugging I realised this didn't work as calling RoomDAO queries in a separate thread without the `IO` context lead to a silent failure of the queries.

Subsequently, I injected the cache clearing code in the `makeWikiedit` function right after a successful query was made. This implementation worked well, and I was able to observe the expected result. I have attached a sped-up video of uploading a image to the WikiSandbox Item.


https://github.com/user-attachments/assets/63ec3286-6db2-4f57-9e5d-6ddc560056c7
